lighthouse-core/scripts/lantern/collect/golden.js

```diff
@@ -0,0 +1,149 @@
+/**
+ * @license Copyright 2019 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+/** @typedef {import('./common.js').Result} Result */
+/** @typedef {import('./common.js').Summary} Summary */
+/** @typedef {import('../run-on-all-assets.js').Golden} Golden */
+
+const fs = require('fs');
+const rimraf = require('rimraf');
+const common = require('./common.js');
+
+/**
+ * @param {LH.Result} lhr
+ * @return {import('../../../audits/metrics.js').UberMetricsItem}
+ */
+function getMetrics(lhr) {
+  const metricsDetails = /** @type {LH.Audit.Details.DebugData=} */ (
+    lhr.audits['metrics'].details);
+  /** @type {import('../../../audits/metrics.js').UberMetricsItem} */
+  return metricsDetails && metricsDetails.items && metricsDetails.items[0];
+}
+
+/**
+ * @template T
+ * @param {T[]} values
+ * @param {(sortValue: T) => number} mapper
+ */
+function getMedianBy(values, mapper) {
+  const resultsWithValue = values.map(value => {
+    return {sortValue: mapper(value), value};
+  });
+
+  resultsWithValue.sort((a, b) => a.sortValue - b.sortValue);
+
+  if (resultsWithValue.length % 2 === 1) {
+    return resultsWithValue[Math.floor(resultsWithValue.length / 2)].value;
+  }
+
+  // Select the value that is closest to the mean.
+  const sum = resultsWithValue.reduce((acc, cur) => acc + cur.sortValue, 0);
+  const mean = sum / resultsWithValue.length;
+  const a = resultsWithValue[Math.floor(resultsWithValue.length / 2)];
+  const b = resultsWithValue[Math.floor(resultsWithValue.length / 2) + 1];
+  const comparison = Math.abs(a.sortValue - mean) < Math.abs(b.sortValue - mean);
+  return comparison ? a.value : b.value;
+}
+
+/**
+ * Returns run w/ the median TTI.
+ * @param {string} url
+ * @param {Result[]} results
+ */
+function getMedianResult(url, results) {
+  // Runs can be missing metrics.
+  const resultsWithMetrics = results.map(result => {
+    const metrics = getMetrics(loadLhr(result.lhr));
+    return {result, metrics};
+  }).filter(({metrics}) => {
+    return metrics && metrics.interactive;
```

Comment:
Should runs without a TTI be considered failed during collection and retried?

Reply:
yeah probably. if it's a legit site that never reaches CPU quiet on a phone then one might argue we should treat it as `Infinity` with infinite error but that's obviously never, ever going to be predicted from a desktop trace. whether legit TTI error or not,  it's not all that useful to have in our test set so we should redo and/or remove the site

Reply:
done